src/linker/Linker/DynamicDependency.cs

```diff
@@ -0,0 +1,140 @@
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Mono.Cecil;
+
+#nullable enable
+
+namespace Mono.Linker
+{
+	/// Tracks dependencies created via DynamicDependencyAttribute in the linker.
+	/// This is almost identical to DynamicDependencyAttribute, but it holds a
+	/// TypeReference instead of a Type, and it has a reference to the original
+	/// CustomAttribute for dependency tracing. It is also a place for helper
+	/// methods related to the attribute.
+	internal class DynamicDependency : Attribute
+	{
+		public CustomAttribute? OriginalAttribute { get; set; }
+		public DynamicDependency (string memberSignature)
+		{
+			MemberSignature = memberSignature;
+		}
+
+		public DynamicDependency (string memberSignature, TypeReference type)
+		{
+			MemberSignature = memberSignature;
+			Type = type;
+		}
+
+		public DynamicDependency (string memberSignature, string typeName, string assemblyName)
+		{
+			MemberSignature = memberSignature;
+			TypeName = typeName;
+			AssemblyName = assemblyName;
+		}
+
+		public DynamicDependency (DynamicallyAccessedMemberTypes memberTypes, TypeReference type)
+		{
+			MemberTypes = memberTypes;
+			Type = type;
+		}
+
+		public DynamicDependency (DynamicallyAccessedMemberTypes memberTypes, string typeName, string assemblyName)
+		{
+			MemberTypes = memberTypes;
+			TypeName = typeName;
+			AssemblyName = assemblyName;
+		}
+
+		public string? MemberSignature { get; }
+
+		public DynamicallyAccessedMemberTypes MemberTypes { get; }
+
+		public TypeReference? Type { get; }
+
+		public string? TypeName { get; }
+
+		public string? AssemblyName { get; }
+
+		public string? Condition { get; set; }
+
+		public static DynamicDependency? ProcessAttribute (LinkContext context, ICustomAttributeProvider provider, CustomAttribute customAttribute)
+		{
+			if (!(provider is MethodDefinition || provider is FieldDefinition))
+				return null;
+
+			if (!DynamicDependency.ShouldProcess (context, customAttribute))
+				return null;
+
+			var dynamicDependency = GetDynamicDependency (context, customAttribute);
+			if (dynamicDependency != null)
+				return dynamicDependency;
+
+			context.LogMessage (MessageContainer.CreateWarningMessage ($"Invalid DynamicDependencyAttribute on '{provider}'", 2030));
+			return null;
+		}
+
+		static DynamicDependency? GetDynamicDependency (LinkContext context, CustomAttribute ca)
+		{
+			var args = ca.ConstructorArguments;
+			if (args.Count < 1 || args.Count > 3)
+				return null;
+
+			// First argument is string or DynamicallyAccessedMemberTypes
+			string? memberSignature = args[0].Value as string;
+			if (args.Count == 1)
+				return memberSignature == null ? null : new DynamicDependency (memberSignature);
+			DynamicallyAccessedMemberTypes? memberTypes = null;
+			if (memberSignature == null) {
+				var argType = args[0].Type;
+				if (!(argType.Namespace == "System.Diagnostics.CodeAnalysis" && argType.Name == "DynamicallyAccessedMemberTypes"))
```

Comment:
Use `IsTypeOf`

Reply:
I will rename the `IsAttribute<T>` helper to `IsTypeOf<T>` and use that in both places.